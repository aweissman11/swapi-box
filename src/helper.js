export const fetchCall = async (url) => {
  const response = await fetch(url);
  if (response.ok) {
    const responseToJSON = await response.json();
    return responseToJSON;
  } else {
    throw new Error('Status is not okay');
  }
};

export const randomNumber = () => {
  const rndm = Math.floor(Math.random() * Math.floor(6));
  return rndm;
};

export const getMovieText = async () => {
  const url = 'https://swapi.co/api/films/';
  //const uncleanMovies = await fetchCall(url);
  const response = await fetch(url)
  const uncleanMovies = await response.json()
  const movies = await cleanMovieText(uncleanMovies.results);
  return movies[randomNumber()];
};

export const cleanMovieText = (uncleanMovies) => {
  const cleanMovies = uncleanMovies.map(movie => {
    return ({title: movie.title, date: movie.release_date, opening: movie.opening_crawl});
  });
  return cleanMovies;
};

export const getItemList = async (page, pageNumber = 1) => {
  const fetchInfo = `${page}/?page=${pageNumber}`;
  const url = `https://swapi.co/api/${fetchInfo}`;
  const uncleanItemList = await fetchCall(url);
  const itemList = await cleanItemList(page, uncleanItemList);
  return itemList;
};

export const cleanItemList = async (page, uncleanItemList) => {
  let cleanList;
  switch (page) {
    case ('people') :
      cleanList = await cleanPeople(uncleanItemList.results);
      return cleanList;
    case ('vehicles') :
      cleanList = await cleanVehicles(uncleanItemList.results);
      return cleanList;
    case ('planets') :
      cleanList = await cleanPlanets(uncleanItemList.results);
      return cleanList;
    default:
      return [];
  }
};

export const cleanVehicles = async (uncleanVehicles) => {
  const unresolvedPromises = await uncleanVehicles.map((vehicle) => {
    return ({
      Name: vehicle.name,
      Model: vehicle.model,
      Class: vehicle.class,
      NumPassengers: vehicle.passengers
    });
  });
  return Promise.all(unresolvedPromises);
};

export const cleanPlanets = (uncleanPlanets) => {
  const unresolvedPromises = uncleanPlanets.map( async (planet) => {
    const residents = await getResidents(planet.residents);
    return ({
      Name: planet.name,
      Terrain: planet.terrain,
      Population: planet.population,
      Climate: planet.climate,
      Residents: residents
    });
  });
  
  return Promise.all(unresolvedPromises);
};

export const getResidents = async (residents) => {
  if (residents.length) {
    const unresolvedPromises = residents.map( async (resident) => {
      const person = await fetchCall(resident);
      return person.name;
    });
  
    return Promise.all(unresolvedPromises);
  } else {
    return Promise.resolve('Nobody lives here');
  }
};

export const cleanPeople = async (uncleanPeople) => {
  const withSpecies = await getSpecies(uncleanPeople);
  const withWorlds = getWorlds(withSpecies);
  return withWorlds;
};

export const getSpecies = (uncleanPeople) => {
  const speciesPromises = uncleanPeople.map( async (person) => {
    const species = await fetchCall(...person.species);
  
    return ({
      ...person,
      Species: species.name
    });
  });
  return Promise.all(speciesPromises);
};

export const getWorlds = (withSpecies) => {
  const worldPromises = withSpecies.map( async (person) => {
    const homeworld = await fetchCall(person.homeworld);
  
    return ({
      Name: person.name,
      Species: person.Species,
      Homeworld: homeworld.name,
      Population: homeworld.population
    });
  });
  return Promise.all(worldPromises);
};
