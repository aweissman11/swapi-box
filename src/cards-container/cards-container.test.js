import React from 'react';
import { shallow, mount } from 'enzyme';

import CardsContainer from './cards-container';
import * as Helper from '../helper';

describe('CardsContainer', () => {

  let currentPage;
  let handleFavorites;
  let wrapper;
  let mockList;
   
  beforeEach(() => {
    handleFavorites = jest.fn();
    currentPage = 'people';

    wrapper = shallow(<CardsContainer 
      getItemList={Helper.getItemList}
      currentPage={currentPage}
      handleFavorites={handleFavorites}
    />);

    mockList = [
      {
        Name: 'Luke',
        Species: 'Human',
        Homeworld: 'Earth',
        Population: 7000000000
      },
      {
        Name: 'Jimbo',
        Species: 'Dinosaur',
        Homeworld: 'Earth',
        Population: 56789
      }
    ];

    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({ 
      ok: true, json: () => Promise.resolve(mockList)
    }));
  });
  
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  it('should start with pageRepo in state', () => {
    wrapper = mount(<CardsContainer 
      getItemList={Helper.getItemList}
      currentPage={currentPage}
      handleFavorites={handleFavorites}
    />);
    expect(wrapper.instance().state.pageRepo).toEqual([]);
  });
  
  it('should show all ten items', () => {
    wrapper.setState({
      pageRepo: Helper.getItemList('people')
    });

    console.log(wrapper.state);
    
    
    
    
    // wrapper = mount(<CardsContainer 
    //   currentPage={currentPage}
    //   getItemList={Helper.getItemList}
    //   handleFavorites={handleFavorites}
    // />);
      
    // wrapper.update();
    // console.log(wrapper.instance().state);
      
    // wrapper = mount(<CardsContainer 
    //   currentPage='planets' 
    //   getItemList={Helper.getItemList}
    //   handleFavorites={handleFavorites}
    // />);
    // wrapper.update();
    // console.log(wrapper.instance().state);
  });
}); 

