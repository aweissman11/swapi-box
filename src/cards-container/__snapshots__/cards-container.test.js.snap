// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardsContainer
    currentPage="people"
    getItemList={[Function]}
    handleFavorites={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="cards-container"
      >
        <div
          className="container"
        >
          <div>
            <div
              className="cards"
            >
              <h1>
                 Repo Loading... 
              </h1>
            </div>
          </div>
        </div>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="container"
        >
          <div>
            <div
              className="cards"
            >
              <h1>
                 Repo Loading... 
              </h1>
            </div>
          </div>
        </div>,
        "className": "cards-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            <div
              className="cards"
            >
              <h1>
                 Repo Loading... 
              </h1>
            </div>
          </div>,
          "className": "container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="cards"
            >
              <h1>
                 Repo Loading... 
              </h1>
            </div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h1>
                 Repo Loading... 
              </h1>,
              "className": "cards",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " Repo Loading... ",
              },
              "ref": null,
              "rendered": " Repo Loading... ",
              "type": "h1",
            },
            "type": "div",
          },
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="cards-container"
        >
          <div
            className="container"
          >
            <div>
              <div
                className="cards"
              >
                <h1>
                   Repo Loading... 
                </h1>
              </div>
            </div>
          </div>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="container"
          >
            <div>
              <div
                className="cards"
              >
                <h1>
                   Repo Loading... 
                </h1>
              </div>
            </div>
          </div>,
          "className": "cards-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              <div
                className="cards"
              >
                <h1>
                   Repo Loading... 
                </h1>
              </div>
            </div>,
            "className": "container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="cards"
              >
                <h1>
                   Repo Loading... 
                </h1>
              </div>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h1>
                   Repo Loading... 
                </h1>,
                "className": "cards",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": " Repo Loading... ",
                },
                "ref": null,
                "rendered": " Repo Loading... ",
                "type": "h1",
              },
              "type": "div",
            },
            "type": "div",
          },
          "type": "div",
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
